extensions-core/protobuf-extensions/src/main/java/org/apache/druid/data/input/protobuf/ProtobufConverter.java

```diff
@@ -0,0 +1,245 @@
+  {
+    if (msg == null) {
+      return null;
+    }
+    final Map<Descriptors.FieldDescriptor, Object> fields = msg.getAllFields();
+    final Map<String, Object> converted = Maps.newHashMapWithExpectedSize(fields.size());
+    for (Map.Entry<Descriptors.FieldDescriptor, Object> field : fields.entrySet()) {
+      converted.put(field.getKey().getJsonName(), convertField(field.getKey(), field.getValue()));
+    }
+    return converted;
+  }
+
+  @Nullable
+  private static Object convertField(Descriptors.FieldDescriptor field, Object value)
+      throws InvalidProtocolBufferException
+  {
+    // handle special types
+    if (value instanceof Message) {
+      Message msg = (Message) value;
+      final String typeName = msg.getDescriptorForType().getFullName();
+      SpecializedConverter converter = SPECIAL_CONVERSIONS.get(typeName);
+      if (converter != null) {
+        return converter.convert(msg);
+      }
+    }
+
+    if (field.isMapField()) {
+      return convertMap(field, value);
+    } else if (field.isRepeated()) {
+      return convertList(field, (List<?>) value);
+    } else {
+      return convertSingleValue(field, value);
+    }
+  }
+
+  @Nonnull
+  private static List<Object> convertList(Descriptors.FieldDescriptor field, List<?> value)
+      throws InvalidProtocolBufferException
+  {
+    final List<Object> theList = Lists.newArrayListWithExpectedSize(value.size());
+    for (Object element : value) {
+      theList.add(convertSingleValue(field, element));
+    }
+    return theList;
+  }
+
+  @Nullable
+  private static Object convertMap(Descriptors.FieldDescriptor field, Object value)
+      throws InvalidProtocolBufferException
+  {
+    final Descriptors.Descriptor type = field.getMessageType();
+    final Descriptors.FieldDescriptor keyField = type.findFieldByName("key");
+    final Descriptors.FieldDescriptor valueField = type.findFieldByName("value");
+    if (keyField == null || valueField == null) {
+      throw new InvalidProtocolBufferException("Invalid map field.");
+    }
+
+    @SuppressWarnings("unchecked")
+    final List<Object> elements = (List<Object>) value;
+    final HashMap<String, Object> theMap = Maps.newHashMapWithExpectedSize(elements.size());
+    for (Object element : elements) {
+      Message entry = (Message) element;
+      theMap.put(
+          (String) convertSingleValue(keyField, entry.getField(keyField)),
+          convertSingleValue(valueField, entry.getField(valueField))
+      );
+    }
+    return theMap;
+  }
+
+  @Nullable
+  private static Object convertSingleValue(Descriptors.FieldDescriptor field, Object value)
+      throws InvalidProtocolBufferException
+  {
+    switch (field.getType()) {
+      case BYTES:
+        return ((ByteString) value).toByteArray();
+      case ENUM:
+        // Special-case google.protobuf.NullValue (it's an Enum).
+        if (field.getEnumType().getFullName().equals("google.protobuf.NullValue")) {
```

Comment:
Is String equality really the best way?

If it is, do it as `"google.protobuf...".equals()` instead.

Reply:
not sure if this is best or not, this is what `JsonFormat` was doing which I adapted most of this code form. I can switch the order though on next PR since I think i'm going to have to add some test coverage anyway to make 🤖 happy